src/waiter.c

```diff
@@ -0,0 +1,311 @@
+                   waiter, data, notify_complete_callback, context);
+    }
+    else
+    {
+        SM_RESULT sm_exec_begin_result = sm_exec_begin(waiter->sm);
+        if (sm_exec_begin_result != SM_EXEC_GRANTED)
+        {
+            LogError("Failure in sm_exec_begin(waiter->sm=%p)", waiter->sm);
+        }
+        else
+        {
+            bool success;
+            srw_lock_acquire_exclusive(waiter->lock);
+            {
+                if (waiter->notified)
+                {
+                    LogError("waiter_notify called twice on WAITER waiter=%p."
+                        "Previous arguments: THANDLE(RC_PTR) data=%" PRI_RC_PTR ", NOTIFY_COMPLETE_CALLBACK notify_complete_callback=%p, void* context=%p."
+                        "Current arguments: THANDLE(RC_PTR) data=%" PRI_RC_PTR ", NOTIFY_COMPLETE_CALLBACK notify_complete_callback=%p, void* context=%p.",
+                        waiter,
+                        waiter->data, waiter->notify_complete_callback, waiter->notify_complete_callback_context,
+                        data, notify_complete_callback, context
+                    );
+                    success = false;
+                }
+                else
+                {
+                    THANDLE(ASYNC_OP) async_op = async_op_create(cancel_notify, sizeof(WAITER_HANDLE), alignof(WAITER_HANDLE), dispose);
+                    if (async_op == NULL)
+                    {
+                        LogError("Failure in async_op_create(cancel_notify=%p, sizeof(WAITER_HANDLE)=%" PRIu32 ", alignof(WAITER_HANDLE)=%" PRIu32 ", dispose=%p)", cancel_notify, (uint32_t)sizeof(WAITER_HANDLE), (uint32_t)alignof(WAITER_HANDLE), dispose);
+                        success = false;
+                    }
+                    else
+                    {
+                        ((ASYNC_OP*)async_op)->context = waiter;
+
+                        THANDLE_ASSIGN(RC_PTR)(&waiter->data, data);
+                        waiter->notify_complete_callback = notify_complete_callback;
+                        waiter->notify_complete_callback_context = context;
+                        waiter->notified = true;
+                        try_complete(waiter);
+                        THANDLE_INITIALIZE_MOVE(ASYNC_OP)(&result, &async_op);
+                        success = true;
+                    }
+                }
+            }
+            srw_lock_release_exclusive(waiter->lock);
+            if (!success)
+            {
+                LogError("Failure in waiter_notify(waiter=%p, data=%" PRI_RC_PTR ", notify_complete_callback=%p, context=%p)", waiter, data, notify_complete_callback, context);
+            }
+            else
+            {
+                goto all_ok;
+            }
+            sm_exec_end(waiter->sm);
+        }
+    }
+all_ok:
+    return result;
+}
```

Comment:
waiting because
1) curiosity
2) I think too many ASYNC_OPs are created
3) reentrancy is not fully explained/tested
4) sm+lock != love. So: threading model.